packages/sitecore-jss-react/src/enhancers/usePersonalization.tsx

```diff
@@ -0,0 +1,91 @@
+import { ComponentRendering, LayoutPersonalizationService } from '@sitecore-jss/sitecore-jss';
+import { useEffect, createElement, useReducer } from 'react';
+import { MissingComponent } from '../components/MissingComponent';
+import { ComponentFactory } from '../components/sharedTypes';
+import { useComponentFactory } from './withComponentFactory';
+
+export interface UsePersonalizationOptions {
+  uid: string;
+  layoutPersonalizationService: LayoutPersonalizationService;
+  missingComponentComponent?: React.ComponentClass<unknown> | React.FC<unknown>;
+}
+
+export interface UsePersonalizationResult {
+  isLoading: boolean;
+  personalizedComponent: React.ReactElement | null;
+}
+
+/**
+ * This hook encapsulates awaiting for personalized component and its creation.
+ * @param {UsePersonalizationOptions} options
+ */
+export function usePersonalization(options: UsePersonalizationOptions): UsePersonalizationResult {
+  // forceUpdate emulating, we need to re-render the component after personalization loading
+  const [, forceUpdate] = useReducer((x) => x + 1, 0);
+  const componentFactory = useComponentFactory();
+
+  const personalizedComponentLayout = options.layoutPersonalizationService.getPersonalizedComponent(
+    options.uid
+  );
+  const isLoading = options.layoutPersonalizationService.isLoading();
+
+  useEffect(() => {
```

Comment:
Not sure that this logic with `forceUpdate` is right. Can we depend on `isLoading` variable?
we can use
```
useEffect(() => {
}, [isLoading])
```
At the top, for example, you can use
```
const [isLoading, setLoading] = useState(options.layoutPersonalizationService.isLoading());
...
useEffect(() => {
   if (!isLoading) return;

   await options.layoutPersonalizationService.loadPersonalizedComponent(options.uid);

   setLoading(false);

}, [isLoading])
```

Reply:
We initially considered such a solution, however, it, unfortunately, has a flaw:
In case of client navigation to the same page (on the `/styleguide` page click `/styleguide` link on the navigation pane), the component is not initializing, but updating. In this case, the state is the same, so `useEffect` will not be run as the state is the same. This makes all design solutions that put loading status into component state inapropriate.